timeseries/blueflood/blueflood.go

```diff
@@ -114,276 +95,228 @@ func NewBlueflood(c Config) timeseries.StorageAPI {
-	go func() {
-		select {
-		case ticket := <-ctx.tickets:
-			series, err := work()
-			// Put the ticket back (regardless of whether caller drops)
-			ctx.tickets <- ticket
-			// Store the result
-			*result = series
-			// Return the error (and sync up with the caller).
-			channel <- err
-		case <-timeout.Done():
-			channel <- timeseries.Error{Code: timeseries.FetchTimeoutError}
+// ChooseResolution will choose the finest-grained resolution for which an interval fetch plan exists that
+// is at least as coarse as the lower bound.
+func (b *Blueflood) ChooseResolution(requested api.Timerange, lowerBound time.Duration) (time.Duration, error) {
+	now := b.config.TimeSource()
+	for i, current := range b.config.Resolutions {
+		if current.Resolution < lowerBound || current.Resolution < requested.Resolution() {
+			continue
+		}
+		_, err := planFetchIntervals(b.config.Resolutions[:i+1], now, requested.Interval())
+		if err == nil {
+			return current.Resolution, nil
 		}
-	}()
-}
-
-func (b *Blueflood) fetchManyLazy(timeout *tasks.Timeout, works []func() (api.Timeseries, error)) ([]api.Timeseries, error) {
-	results := make([]api.Timeseries, len(works))
-	channel := make(chan error, len(works)) // Buffering the channel means the goroutines won't need to wait.
-
-	limit := b.config.MaxSimultaneousRequests
-	tickets := make(chan struct{}, limit)
-	for i := 0; i < limit; i++ {
-		tickets <- struct{}{}
-	}
-	ctx := BluefloodParallelRequest{
-		tickets: tickets,
-	}
-	for i := range results {
-		b.fetchLazy(timeout, &results[i], works[i], channel, ctx)
 	}
+	return 0, fmt.Errorf("cannot choose resolution for timerange %+v; available resolutions do not live long enough or are not available soon enough.", requested)
+}
 
-	var err error
-	for range works {
-		select {
-		case thisErr := <-channel:
-			if thisErr != nil {
-				err = thisErr
+// planFetchIntervals will plan the (point-count minimal) request intervals needed to cover the given timerange.
+// the resolutions slice should be sorted, with the finest-grained resolution first.
+func planFetchIntervals(resolutions []Resolution, now time.Time, requestInterval api.Interval) (map[Resolution]api.Interval, error) {
+	answer := map[Resolution]api.Interval{}
+	// Note: for anything other than FULL, a Blueflood returned point corresponds to the period FOLLOWING that point.
+	// e.g. at 1hr resolution, a 4pm point summarizes all points in [4pm, 5pm], exclusive of 5pm.
+	requestTimerange := requestInterval.CoveringTimerange(resolutions[len(resolutions)-1].Resolution)
+	here := requestTimerange.Start()
+	end := requestTimerange.End()
+	for i := len(resolutions) - 1; i >= 0; i-- {
+		resolution := resolutions[i]
+		if here.Before(now.Add(-resolution.TimeToLive)) {
+			// Expired
+			continue
+		}
+		originalHere := here
+		for {
+			// TODO: optimize this into a division.
```

Comment:
I assume this comment applies to the entire `for`? If so let's move the comment up there.